File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        d = {}

        for num in nums:
            if num not in d:
                d[num] = 0
            d[num]+=1
        temp_heap = []
        result = []
        for key, value in d.items():
            temp_heap.append([value, key])
        heapq.heapify(result)
        for value in temp_heap:
            if len(result)<k:
                heapq.heappush(result, value)
            else:
                if value[0]>result[0][0]:
                    heapq.heapreplace(result, value)
        return [value[1] for value in result]
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prefix = 1;
        int n = nums.size();

        vector<int> result(n, 1);

        for(int i = 0 ; i< n; i++) {
            result[i]*=prefix;
            prefix*=nums[i];
        }

        int postfix = 1;

        for(int j = n - 1; j>=0; j--) {
            result[j]*=postfix;
            postfix*=nums[j];
        }
        
    return result;
    }
};
